Let's Encrypt CI & testing support & fixes, bump retries, update docs

thunderysteak opened the pull request:
Added support for testing Let's Encrypt using a local Pebble server. There's a strange issue with Pebble where its not able to do response challenge when on the same host and it might be an upstream issue, as Pebble server was returning 403 during challenge response. Due of this I made all challenges to be accepted even if they're invalid, but it shouldn't cause issues in production deployment.

Added extra warnings to not blindly enable the use_http variable without configuring Mastodon for it as I saw people blindly turning that on.

This should hopefully fix #43 

Tested only on RHEL9 due to time constrains but it passes all local tests on it. 

```
PLAY RECAP *********************************************************************
rhel9                      : ok=89   changed=29   unreachable=0    failed=0    skipped=52   rescued=0    ignored=0   

==> rhel9: Running provisioner: shell...
    rhel9: Running: inline script
    rhel9: Running PostgreSQL commands required for testing
==> rhel9: Running provisioner: shell...
    rhel9: Running: inline script
    rhel9: Running Goss tests:
    rhel9: The target is rhel
    rhel9:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
    rhel9:                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 11.8M  100 11.8M    0     0  7356k      0  0:00:01  0:00:01 --:--:-- 22.0M
    rhel9: /tmp/goss: OK
    rhel9: .....................................................
    rhel9: 
    rhel9: Total Duration: 0.389s
    rhel9: Count: 53, Failed: 0, Skipped: 0
thunderysteak@Steaks-TrashBook-Pro mastodon-ansible % 
```

thunderysteak commented:
Rocky 8 is failing due of Certbot needing cffi>=1.12 but only cffi 1.11.5 being available. I assume that this is a similar issue on RHEL7 due of old Python.

A possible fix is to just update it through pip, but we shouldn't mix pip and package manager packages. Should be able to bump Python up to version 3.9+

I'll see if I can do something about it during the weekend

oliverlorenz commented:
I run this on ubuntu 22.10 and it fully works! Thanks a lot @thunderysteak!

thunderysteak commented:
I switched the Ansible package installation in Vagrant from package manager based to pip based as per recommendation from [Jeff Geerling @geerlingguy on Twitter](https://twitter.com/geerlingguy/status/1596593622044274688)

We need to use pip3 instead of pip for ansible install mode or we will encounter https://github.com/hashicorp/vagrant/issues/12155 and have focal builds fail. This should fix the intermittent failures of trying to install Ansible in the VMs

thunderysteak commented on bare/roles/web/tasks/letsencrypt.yml:
```diff
@@ -2,30 +2,54 @@
     minute: "15"
     hour: "0"
-    job: "letsencrypt renew && service nginx reload"
+    job: "/opt/certbot/bin/python -c 'import random; import time; time.sleep(random.random() * 3600)' && sudo certbot renew -q && service nginx reload"
```
This is based on Certbot's own documentation:
https://certbot.eff.org/instructions?ws=other&os=pip

thunderysteak replied:
EFF gives this guidance to stagger the renewal by a few random minutes in case the host has multiple letsencrypt certificates that its renewing

thunderysteak commented on goss.yaml:
```diff
@@ -14,6 +14,14 @@ file:
+  /etc/letsencrypt/live/mastodon.local/privkey.pem:
+    exists: true
+    filetype: "symlink"
+    mode: "0777" 
```
This is default Certbot behavior, as such, the goss test tests for default Certbot behavior.

Example Certbot output on RHEL8 system

```
[thunderysteak@webhost ~]$ sudo certbot --nginx -d example.domain
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Renewing an existing certificate for example.domain

Successfully received certificate.
Certificate is saved at: /etc/letsencrypt/live/example.domain/fullchain.pem
Key is saved at:         /etc/letsencrypt/live/example.domain/privkey.pem
This certificate expires on 2023-05-11.
These files will be updated when the certificate renews.
Certbot has set up a scheduled task to automatically renew this certificate in the background.

Deploying certificate
Successfully deployed certificate for example.domain to /etc/nginx/conf.d/mastodon-reverseproxy.conf
Your existing certificate has been successfully renewed, and the new certificate has been installed.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
If you like Certbot, please consider supporting our work by:
 * Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
 * Donating to EFF:                    https://eff.org/donate-le
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
[thunderysteak@webhost ~]$ cat /etc/nginx/conf.d/mastodon-reverseproxy.conf | grep Certbot
    } # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/example.domain/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/example.domain/privkey.pem; # managed by Certbot
[thunderysteak@webhost ~]$ sudo ls -lah /etc/letsencrypt/live/example.domain/
total 4.0K
drwxr-xr-x. 2 root root  93 Feb 10 18:54 .
drwx------. 5 root root  89 May 14  2022 ..
lrwxrwxrwx. 1 root root  35 Feb 10 18:54 cert.pem -> ../../archive/example.domain/cert4.pem
lrwxrwxrwx. 1 root root  36 Feb 10 18:54 chain.pem -> ../../archive/example.domain/chain4.pem
lrwxrwxrwx. 1 root root  40 Feb 10 18:54 fullchain.pem -> ../../archive/example.domain/fullchain4.pem
lrwxrwxrwx. 1 root root  38 Feb 10 18:54 privkey.pem -> ../../archive/example.domain/privkey4.pem
-rw-r--r--. 1 root root 692 May 14  2022 README

```

moritzheiber commented:
@thunderysteak Hey, sorry for being away from this PR for a while. Would you mind rebasing it and maybe addressing a few of the comments? I'd love to get it merged now that Vagrant supports installing the right version of Ansible.

thunderysteak commented:
@moritzheiber will do, might take a white as I'm rebuilding my dev setup from scratch

thunderysteak commented:
Apologies @moritzheiber this had to be further delayed as I'm still recovering from COVID

moritzheiber commented:
Don’t worry about it. Your health and well-being takes absolute precedence over shuffling around a few YAML files. I wish you a speedy recovery!

thunderysteak commented:
@moritzheiber Requested changes should be fixed now.

* Removed the debug strings from `Vagrantfile`
* Email address is no longer prefilled for Let's Encrypt and updated the doc and template to reflect this
* There's now a preflight check to check if an email address was filled out
* Auto-update of certbot is now behind a feature flag that is disabled by default
* Reworked the installation of certbot due of Python 3.6 becoming EoL & Ubuntu oddities (Ubuntu doesn't want to accept venv installed via PIP,...)
* Added docs about `letsencrypt.yml` as it was missing
* Goss now dynamically checks the Ruby version installed against the Mastodon installed via `.ruby-version`. Another Ubuntu oddity here where only Ubuntu converts `true` & `false` strings into booleans and breaking Goss, so alt wording had to be used.
* New way of installing certbot is behind a feature flag to prevent breakages against older versions of playbook. User has to explicitly enable it. Noted in doc.
* Removed Ansible installation via pip3 during rebase

moritzheiber commented:
@thunderysteak Thank you again for your hard work. Unfortunately, I think you've pulled in a lot of existing commits which are now showing up as separate changes (for some reason) and therefore this PR isn't mergeable at all :disappointed:

Maybe re-creating the PR will solve this?

thunderysteak commented:
@moritzheiber I think something went wrong during the rebase when I used Gitkraken. Will close this and recreate the PR
